Add EIP: ModExp Gas Cost Increase (#9356)
ethereum/EIPs: EIPS/eip-7883.md

---
eip: 7883
title: ModExp Gas Cost Increase
description: Increases cost of ModExp precompile
author: Marcin Sobczak (@marcindsobczak), Marek Moraczyński (@MarekM25), Marcos Maceo (@stdevMac)

> discussions-to: TBD

Reviewer: ```suggestion
discussions-to: https://ethereum-magicians.org/t/eip-7883-modexp-gas-cost-increase/22841
```

status: Draft
type: Standards Track
category: Core
created: 2025-02-11
requires: 2565
---

## Abstract

> This EIP is modifying the `ModExp` pricing algorithm introduced in [EIP-2565](./eip-2565.md).

Reviewer: ```suggestion
This EIP is modifying the `ModExp` precompile pricing algorithm introduced in [EIP-2565](./eip-2565.md).
```

## Motivation

> There are underpriced `ModExp` scenarios for it's resource consumption. By modifying `ModExp` pricing formula, these scenarios are covered with minimal impact on real world usages. Target is to make all `ModExp` scenarios at least as fast as EcRecover precompile.

Reviewer: ```suggestion
There are cases where the `ModExp` precompile is underpriced for it's resource consumption. By modifying the `ModExp` pricing formula these scenarios would be covered with minimal impact on real world applications. The target is to make `ModExp` at least as fast as EcRecover precompile in all cases.
```

> 

Reviewer: Could mention mitigating worst cases can help with increasing the gas limit safely

## Specification

> As of `FORK_BLOCK_NUMBER`, the gas cost of calling the precompile at address `0x0000000000000000000000000000000000000005` will be calculated as follows:

Reviewer 2: we don't refer hardfork number of timestamp now, you may say upon activation of this EIP, also forks are now timestamp based

Author: Copy-pasted this part brainless from EIP-2565 (last modification od ModExp pricing)

This part of equation:

```
    return max(200, math.floor(multiplication_complexity * iteration_count / 3))
```

Is replaced by this:

```
    return max(500, math.floor(multiplication_complexity * iteration_count / 3))
```

### 2. Increase cost when exponent is larger than 32 bytes

This part of equation:

```
    elif exponent_length > 32: iteration_count = (8 * (exponent_length - 32)) + ((exponent & (2**256 - 1)).bit_length() - 1)
```

Is replaced by this:

```
    elif exponent_length > 32: iteration_count = (16 * (exponent_length - 32)) + ((exponent & (2**256 - 1)).bit_length() - 1)
```

Multiplier 8 is replaced by 16.

### 3. Increase cost when base or modulus is larger than 32 bytes

This part of equation:

```
def calculate_multiplication_complexity(base_length, modulus_length):
    max_length = max(base_length, modulus_length)
    words = math.ceil(max_length / 8)
    return words**2
```

Is replaced by this:

```
def calculate_multiplication_complexity(base_length, modulus_length):
    max_length = max(base_length, modulus_length)
    words = math.ceil(max_length / 8)
    multiplication_complexity = 0
    if max_length <= 32: multiplication_complexity = words**2

>     elif max_length > 32: multiplication_complexity = 2 * words**2

Reviewer 3: To make this logic more cleaner I will suggest this 
```suggestion
    else:
        multiplication_complexity = 2 * words ** 2
```

Author: I modified pricing formula introduced in [EIP-2565](https://eips.ethereum.org/EIPS/eip-2565) and for consistency EIP-7883 is using the same convention

Reviewer 3: Okay

| modexp_nagydani_3_qube       | 341 | 682 | 100% |
| modexp_nagydani_3_pow0x10001 | 5461 | 10922 | 100% |
| modexp_nagydani_4_square     | 1365 | 2730 | 100% |
| modexp_nagydani_4_qube       | 1365 | 2730 | 100% |
| modexp_nagydani_4_pow0x10001 | 21845 | 43690 | 100% |
| modexp_nagydani_5_square     | 5461 | 10922 | 100% |
| modexp_nagydani_5_qube       | 5461 | 10922 | 100% |
| modexp_nagydani_5_pow0x10001 | 87381 | 174762 | 100% |
| modexp_marius_1_even         | 2057 | 3774 | 83% |
| modexp_guido_1_even          | 2298 | 4261 | 85% |
| modexp_guido_2_even          | 2300 | 4262 | 85% |
| modexp_guido_3_even          | 5400 | 10800 | 100% |
| modexp_guido_4_even          | 1026 | 1967 | 92% |
| modexp_marcin_1_base_heavy   | 200 | 500 | 150% |
| modexp_marcin_1_exp_heavy    | 215 | 500 | 133% |
| modexp_marcin_1_balanced     | 200 | 500 | 150% |
| modexp_marcin_2_base_heavy   | 867 | 1734 | 100% |
| modexp_marcin_2_exp_heavy    | 852 | 1364 | 60% |
| modexp_marcin_2_balanced     | 996 | 1992 | 100% |
| modexp_marcin_3_base_heavy   | 677 | 677 | 0% |
| modexp_marcin_3_exp_heavy    | 765 | 765 | 0% |
| modexp_marcin_3_balanced     | 1360 | 1360 | 0% |

> ## Reference Implementation

Reviewer: This section is optional so can be removed
